### PR TITLE
EVM: optimize stack

### PR DESCRIPTION
**Pull request by jochem-brouwer**
This optimizes stack.

In particular, our previous stack kept deleting / adding items to stack. Here, the internal stack Array will only grow. 

There are two error checks removed from `stack.push`: since we only call the stack internally, we are sure: 

- The argument is always of type `bigint`
- The argument is always within `2**256-1` (`uint256`)
- Therefore, we don't have to do these error checks.

I have tested initializing the Stack with 1024 empty items, this makes the constructor 10x slower and the gains are not that good. On top of the optimizations of #2996 this optimizes the performance JUMPDEST tests from ~8.1 MGas/s to 8.5 MGas/s. (This is the test which uses this code: `0x60025B8056` -> just infinitely keep JUMPing to a JUMPDEST)

Note: target branch is also the PR of #2996.

To profile the new stack, follow the instructions in this gist: https://gist.github.com/jochem-brouwer/3969933a0eb6bb4861c1ef4e61ed3f70

Output from my side:

```
push on empty stack: Diff (neg is better): -0.377s (-41.08%)
push on almost-full-stack: Diff (neg is better): -0.261s (-29.95%)
pop on empty stack: Diff (neg is better): -0.021s (-3.48%)
pop on full-stack: Diff (neg is better): -0.028s (-4.97%)
popN 1 on stack size 1: Diff (neg is better): -0.386s (-39.98%)
popN 1 on full-stack: Diff (neg is better): -0.398s (-38.01%)
popN 8 on stack size 8: Diff (neg is better): -0.383s (-32.91%)
popN 8 on full-stack: Diff (neg is better): -0.504s (-38.91%)
peek 1 on stack size 1: Diff (neg is better): -0.136s (-17.6%)
peek 1 on full-stack: Diff (neg is better): -0.155s (-19.25%)
peek 8 on stack size 8: Diff (neg is better): -0.066s (-7.58%)
peek 8 on full-stack: Diff (neg is better): -0.065s (-7.44%)
swap 1 on stack size 2: Diff (neg is better): 0.009s (1.48%)
swap 1 on full-stack: Diff (neg is better): 0.024s (4.3%)
swap 16 on stack size 17: Diff (neg is better): 0.017s (3.07%)
swap 16 on full-stack: Diff (neg is better): 0.015s (2.61%)
dup 1 on stack size 1: Diff (neg is better): -0.449s (-45.2%)
dup 1 on almost-full-stack: Diff (neg is better): -0.428s (-42.38%)
dup 16 on stack size 16: Diff (neg is better): -0.455s (-44.69%)
dup 16 on almost-full-stack: Diff (neg is better): -0.448s (-43.78%)
```

JUMPDEST test:

```
gas: 99979000, time: 11.483665864944458, Mgas/s: 8.70619201009673
gas: 99979000, time: 11.755131164073944, Mgas/s: 8.505136914639968
gas: 99979000, time: 11.642706596970559, Mgas/s: 8.587264410323163
gas: 99979000, time: 11.971259549975395, Mgas/s: 8.351585694272705
```

Sample: 

```
Profile: JUMPDEST loop (gas: 100000000)
+== Opcodes performance - Tx 0x4766bc7b95062417f71bcd725f5331989f0c3ecc4ab75d4db52674423d699aa6 ==+
|---------------------------------------------------------------|
|      tag |   calls | ms/call | total (ms) | gas used | Mgas/s |
|     JUMP | 8331583 |       0 |   2676.627 | 66652664 | 24.902 |
|      DUP | 8331584 |       0 |   1883.657 | 24994752 | 13.269 |
| JUMPDEST | 8331584 |       0 |    1832.62 |  8331584 |  4.546 |
|     PUSH |       1 |   0.172 |      0.172 |        3 |  0.017 |
+---------------------------------------------------------------+
gas: 99979000, time: 11.971259549975395, Mgas/s: 8.351585694272705
```

### CODE REVIEW

File: packages/evm/src/stack.ts [hunk 1]
```diff
@@ -83,16 +83,17 @@ export class Stack {
...
-    if (this._store.length <= position) {
+    if (this._len <= position) {
       throw new EvmError(ERROR.STACK_UNDERFLOW)
     }
 
-    const head = this._store.length - 1
-    const i = this._store.length - position - 1
+    const head = this._len - 1
+    const i = head - position
+    const sto = this._store
```
Comment (acolytec3) [inline packages/evm/src/stack.ts:92]: Is `sto` just short for `store`?  It wasn't obvious to me at first.

**Review by acolytec3 — CHANGES_REQUESTED** [review]
Looks great so far but would be curious if you could look at my couple of suggested alternative implementations and compare their performance to what you currently have coded.

**Comment by jochem-brouwer** [comment]
Same version as used as results in the opening post of this PR, but now the results with initializing the stack array to 1024 items:

```
push on empty stack: Diff (neg is better): -0.321s (-36.47%)
push on almost-full-stack: Diff (neg is better): -0.255s (-29.23%)
pop on empty stack: Diff (neg is better): 0.003s (0.4%)
pop on full-stack: Diff (neg is better): -0.033s (-5.63%)
popN 1 on stack size 1: Diff (neg is better): -0.389s (-40.18%)
popN 1 on full-stack: Diff (neg is better): -0.438s (-40.6%)
popN 8 on stack size 8: Diff (neg is better): -0.398s (-33.34%)
popN 8 on full-stack: Diff (neg is better): -0.498s (-38.79%)
peek 1 on stack size 1: Diff (neg is better): -0.16s (-20.83%)
peek 1 on full-stack: Diff (neg is better): -0.094s (-12.09%)
peek 8 on stack size 8: Diff (neg is better): 0.009s (0.99%)
peek 8 on full-stack: Diff (neg is better): 0.027s (3.12%)
swap 1 on stack size 2: Diff (neg is better): 0.005s (0.8%)
swap 1 on full-stack: Diff (neg is better): 0.01s (1.73%)
swap 16 on stack size 17: Diff (neg is better): 0.013s (2.33%)
swap 16 on full-stack: Diff (neg is better): 0.02s (3.5%)
dup 1 on stack size 1: Diff (neg is better): -0.426s (-42.42%)
dup 1 on almost-full-stack: Diff (neg is better): -0.441s (-42.55%)
dup 16 on stack size 16: Diff (neg is better): -0.442s (-42.58%)
dup 16 on almost-full-stack: Diff (neg is better): -0.43s (-41.41%)
```

I don't really see a big difference here :thinking:

File: packages/evm/src/stack.ts [hunk 2]
```diff
@@ -9,38 +7,32 @@ export class Stack {
   _store: bigint[]
   _maxHeight: number
 
+  private _len: number = 0
+
   constructor(maxHeight?: number) {
-    this._store = []
+    this._store = [] //Array(maxHeight ?? 1024) -> initializing the array makes the constructor 10x slower
```
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:13]: Stack only gets initialized on creating a `new Interpreter`. This only happens if you enter a new "call frame", so either: the transition from tx -> calling / creating a contract, or using opcodes CALL, CALLCODE, DELEGATECALL, STATICCALL, CREATE, CREATE2. (In practice this thus does not happen a lot since there is a rather big base gas cost for these opcodes)

[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:13]: Note: with TransientStorage, each call frame gets a extra stack (so for Cancun the amount of times the constructor gets called doubles when compared <= Cancun)

File: packages/evm/src/stack.ts [hunk 3]
```diff
@@ -9,38 +7,32 @@ export class Stack {
...
     this._maxHeight = maxHeight ?? 1024
   }
 
   get length() {
-    return this._store.length
+    return this._len
   }
 
   push(value: bigint) {
```
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:21]: We only use this Stack internally, plus we use typescript, so there is no reason to assume that the input is not a bigint.

File: packages/evm/src/stack.ts [hunk 4]
```diff
@@ -49,15 +41,22 @@ export class Stack {
...
     if (num === 0) {
       return []
     }
 
-    return this._store.splice(-1 * num).reverse()
+    const arr = Array(num)
+    const cache = this._store
+
+    for (let pop = 0; pop < num; pop++) {
+      arr[pop] = cache[--this._len]
```
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:56]: Interesting, I have not checked, I honestly assumed that this would be faster instead of using `slice`. This should definitely be checked :)

File: packages/evm/src/stack.ts [hunk 5]
```diff
@@ -66,14 +65,15 @@ export class Stack {
    * @throws {@link ERROR.STACK_UNDERFLOW}
    */
   peek(num: number = 1): bigint[] {
-    const peekArray: bigint[] = []
+    const peekArray: bigint[] = Array(num)
+    let start = this._len
 
-    for (let peek = 1; peek <= num; peek++) {
-      const index = this._store.length - peek
+    for (let peek = 0; peek < num; peek++) {
```
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:71]: Have to check :)

File: packages/evm/src/stack.ts (hunk 1 above)
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:92]: Yes, and I am not sure if "caching" things actually improves something, but in my head it should somewhat improve performance (instead of doing key -> value lookup multiple times on an object where you know that obj[key] has not yet changed)

File: packages/evm/src/stack.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:56]: I tested: 

```typescript
  popN(num: number = 1): bigint[] {
    if (this._len < num) {
      throw new EvmError(ERROR.STACK_UNDERFLOW)
    }

    if (num === 0) {
      return []
    }

    const arr = this._store.slice(this._len - num, this._len).reverse()

    this._len -= num

    return arr
  }
  ```
  Results
  
  ```
  popN 1 on stack size 1: Diff (neg is better): 0.249s (38.97%)
popN 1 on full-stack: Diff (neg is better): 0.279s (41.94%)
popN 2 on stack size 2: Diff (neg is better): 0.346s (50.15%)
popN 2 on full-stack: Diff (neg is better): 0.337s (49.35%)
popN 3 on stack size 3: Diff (neg is better): 0.313s (42.46%)
popN 3 on full-stack: Diff (neg is better): 0.27s (34.98%)
popN 4 on stack size 4: Diff (neg is better): 0.299s (37.97%)
popN 4 on full-stack: Diff (neg is better): 0.282s (36.32%)
popN 5 on stack size 5: Diff (neg is better): 0.256s (31.47%)
popN 5 on full-stack: Diff (neg is better): 0.25s (30.86%)
popN 6 on stack size 6: Diff (neg is better): 0.252s (29.89%)
popN 6 on full-stack: Diff (neg is better): 0.238s (27.67%)
popN 7 on stack size 7: Diff (neg is better): 0.257s (29.06%)
popN 7 on full-stack: Diff (neg is better): 0.227s (26%)
popN 8 on stack size 8: Diff (neg is better): 0.214s (23.62%)
popN 8 on full-stack: Diff (neg is better): 0.208s (23%)
```

It is thus slower, but the "slowness" seems to decrease, so for larger arrays it might indeed be faster. But it seems not for small arrays?

File: packages/evm/src/stack.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:71]: See -> https://github.com/ethereumjs/ethereumjs-monorepo/pull/3000#discussion_r1311517087

**Comment by jochem-brouwer** [comment]
EVM tests locally pass... so not sure why CI fails?

File: packages/evm/src/stack.ts (hunk 1 above)
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts]: Have cleared this up

File: packages/evm/src/stack.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (acolytec3) [reply packages/evm/src/stack.ts:64]: Just to be sure I'm tracking, the `diff` is the time difference between running it with your original optimized loop versus `.slice.reverse()`  and the slice.reverse version is .25s slower?

[the remark this answers is not in the document]
Comment (jochem-brouwer) [reply packages/evm/src/stack.ts:64]: Yes

[the remark this answers is not in the document]
Comment (acolytec3) [reply packages/evm/src/stack.ts:64]: Cool, thanks for exploring this.

**Review by acolytec3 — APPROVED** [review]
LGTM! Great work

File: packages/vm/test/api/EIPs/eip-3855.spec.ts [hunk 6]
```diff
@@ -19,14 +19,11 @@ describe('EIP 3541 tests', () => {
...
     vm.evm.events!.on('step', (e: InterpreterStep) => {
-      if (typeof stack !== 'undefined') {
-        assert.fail('should only do PUSH0 once')
-      }
       stack = e.stack
     })
 
     const result = await vm.evm.runCode!({
-      code: hexToBytes('0x5F'),
+      code: hexToBytes('0x5F00'),
```
Comment (jochem-brouwer) [inline packages/vm/test/api/EIPs/eip-3855.spec.ts:26]: Note: this test had to change because before this PR, the `stack` item in this test is a pointer to the actual internal stack of the EVM. Once `step` is called, the stack is empty. Then, PUSH0 is used and thus 0 is pushed on stack. Thus the pointer to stack is now `[0]`. However, with the new change, we make a copy of the stack to represent the actual state of the stack. Thus, at the step and also after the step, it will not change and stay empty.

If we instead use `STOP` (opcode `0x00`) then the stack itself will thus change to the stack *before* using STOP (thus `[0]`), which fixes this test.

Note: this PR will thus make the `step` event **much** slower since we now copy the stack instead of reporting a pointer to it.

Comment (holgerd77) [reply packages/vm/test/api/EIPs/eip-3855.spec.ts:26]: Actually that was/would be a bit my suggestion, that we fall back to the old behavior once someone is listening to the step event and then one time correct the stack to the correct height and then do the respective pushes and pops.

A bit more code to handle. Would this make sense? 🤔

Comment (jochem-brouwer) [reply packages/vm/test/api/EIPs/eip-3855.spec.ts:26]: Yes, this makes sense. So I will introduce `PerformanceStack` and `OldStack` (will think of better names) and will use `OldStack` if we have an event listener.

There is one edge case, what if we are already running EVM and then we start listening to the step event?

Comment (holgerd77) [reply packages/vm/test/api/EIPs/eip-3855.spec.ts:26]: I don't think this will work too separate this into two stacks. This is not an edge case you mentioned, listeners can come and go all the time during runtime.

But I think it should be possible to do in one integrated stack implementation?

Comment (holgerd77) [reply packages/vm/test/api/EIPs/eip-3855.spec.ts:26]: Ok, will nevertheless merge here for now.
